### PR TITLE
Feature/39 replace attack speed with bullet speed

### LINKED ISSUE #39: Replace-attack-speed-with-bullet-speed
## Problem Description

Change the attack speed display feature to a bullet speed display feature.

## Todo List

-  [x] Update the DrawManager to render the current bullet speed on the screen.
-  [x] Adjust the relevant game logic to track and pass the bullet speed value to the HUD.
-  [x] Ensure the bullet speed display updates dynamically as the player’s bullet speed changes.
-  [x] Test the bullet speed display to ensure it integrates smoothly with existing HUD elements, such as score and player lives.

## Feature Description

The current bullet speed will be displayed on the screen.

### PR DESCRIPTION
**Pull request by yunsful**
## What
This PR adds the Replace attack speed with bullet speed. It includes the following changes:
- Changed the attack speed display method to a bullet speed display method in `DrawManagerImpl`.
- Updated `GameScreen` to display bullet speed instead of attack speed.
- adjust attack speed getter to bullet speed getter.

## Why
The display bullet speed feature can monitor bullet speed changes in real-time.

## Related Issue
Closes #39 

## How to Test
1. Start the application locally.
2. Check if the attack speed feature has been changed to the bullet speed feature.
3. Check if the bullet speed value increases when a bullet speed boost item is collected.

## Screenshots
image

### CODE REVIEW

File: src/HUDTeam/DrawManagerImpl.java [hunk 1]
```diff
@@ -20,10 +20,10 @@ public static void drawLevel(final Screen screen, final int level){
         backBufferGraphics.drawString(levelText, xPosition - fontRegularMetrics.stringWidth(levelText) / 2, yPosition); // edit by jesung ko - TeamHUD
     } // Lee Hyun Woo - level
     
-    public static void drawBulletSpeed(final Screen screen, final double bulletSpeed) {
+    public static void drawBulletSpeed(final Screen screen, final int bulletSpeed) {
         backBufferGraphics.setFont(fontRegular);
         backBufferGraphics.setColor(Color.WHITE);
-        String attackSpeedText = String.format("BS : %.2f px/f ", bulletSpeed);
+        String attackSpeedText = String.format("BS : %d px/f ", bulletSpeed);
```
Comment (DongWonLee2) [inline src/HUDTeam/DrawManagerImpl.java]: Looks a bit more intuitive than a float type. Nice!

File: src/entity/Ship.java [hunk 2]
```diff
@@ -198,12 +198,14 @@ public final int getSpeed() {
...
 	 *
-	 * @return Attack speed (bullets per second).
+	 * @return bullet speed (Pixels per frame).
 	 */
-	public final double getAttackSpeed() {
-		return 1000.0 / growth.getShootingDelay();
+	public final double getBulletSpeed() {
+		double speed = growth.getBulletSpeed();
+		if (speed >= 0) return speed;
+		else return speed * (-1);
```
Comment (DongWonLee2) [inline src/entity/Ship.java:208]: Using Math.abs() makes the code simpler. Like this

return Math.abs(growth.getBulletSpeed());

File: src/HUDTeam/DrawManagerImpl.java [hunk 3]
```diff
@@ -22,10 +20,10 @@ public static void drawLevel(final Screen screen, final int level){
         backBufferGraphics.drawString(levelText, xPosition - fontRegularMetrics.stringWidth(levelText) / 2, yPosition); // edit by jesung ko - TeamHUD
     } // Lee Hyun Woo - level
     
-    public static void drawAttackSpeed(final Screen screen, final double attackSpeed) {
+    public static void drawBulletSpeed(final Screen screen, final double bulletSpeed) {
         backBufferGraphics.setFont(fontRegular);
         backBufferGraphics.setColor(Color.WHITE);
-        String attackSpeedText = String.format("AS : %.2f ", attackSpeed);
+        String attackSpeedText = String.format("BS : %.2f px/f ", bulletSpeed);
```
Comment (DongWonLee2) [inline src/HUDTeam/DrawManagerImpl.java]: I wonder why you wrote 'px/f'

**Review by DongWonLee2 — APPROVED** [review]
screenshot7

I confirmed that it works fine, but I left a comment about the improvements and questions, so please refer to it.

File: src/entity/Ship.java (hunk 2 above)
Comment (yunsful) [reply src/entity/Ship.java:208]: It seems that for the sake of code stability and performance, it would be better not to use the abs() function

Comment (kojesung) [reply src/entity/Ship.java:208]: I think the reviewer's opinion is good.

File: src/HUDTeam/DrawManagerImpl.java (hunk 3 above)
Comment (yunsful) [reply src/HUDTeam/DrawManagerImpl.java]: It means pixels per frame

File: src/entity/Ship.java (hunk 2 above)
Comment (DongWonLee2) [reply src/entity/Ship.java:208]: > I think the reviewer's opinion is good.

Noted.

File: src/entity/Ship.java [hunk 4]
```diff
@@ -198,12 +198,14 @@ public final int getSpeed() {
...
 	/**
-	 * Calculates and returns the attack speed in bullets per second.
+	 * Calculates and returns the bullet speed in Pixels per frame.
 	 *
-	 * @return Attack speed (bullets per second).
+	 * @return bullet speed (Pixels per frame).
 	 */
-	public final double getAttackSpeed() {
-		return 1000.0 / growth.getShootingDelay();
+	public final int getBulletSpeed() {
```
Comment (kojesung) [inline src/entity/Ship.java:205]: It seems that author information should be included in the comments.

**Review by kojesung — CHANGES_REQUESTED** [review]

File: src/entity/Ship.java (hunk 2 above)
Comment (sml-logs) [reply src/entity/Ship.java:208]: Why do you think using the abs() function reduces the stability and performance of your code?

File: src/HUDTeam/DrawManagerImpl.java
Comment (sml-logs) [inline src/HUDTeam/DrawManagerImpl.java:1]: Changing the name from attackSpeed ​​to bulletSpeed ​​makes it clearer and less confusing.

**Review by aeioiie — APPROVED** [review]
It's lagging badly, but I don't think it's the problem with this code. LGTM!

https://github.com/user-attachments/assets/27352a21-6d0c-4f4b-a263-d3480c5b5d84

**Review by sml-logs — APPROVED** [review]
I have confirmed that all functions related to this code are working properly.
ScreenShot 2024-10-11 오후 6 00 36

File: src/entity/Ship.java (hunk 2 above)
Comment (yunsful) [reply src/entity/Ship.java:208]: There are certain numbers(MIN_VALUE) for which the abs() function does not work correctly and The regular logic is approximately twice as fast.

File: src/entity/Ship.java (hunk 4 above)
Comment (yunsful) [reply src/entity/Ship.java:205]: It has been reflected.

File: src/HUDTeam/DrawManagerImpl.java
Comment (yunsful) [reply src/HUDTeam/DrawManagerImpl.java:1]: Thank you for confirming.
Please check it.

File: src/entity/Ship.java (hunk 2 above)
Comment (sml-logs) [reply src/entity/Ship.java:208]: Ah, I see. Thank you for your kind explanation.

**Comment by DongWonLee2** [comment]
> It's lagging badly, but I don't think it's the problem with this code. LGTM!
> 
>  review.mp4

I am also aware of the problem situation. It was judged to be a problem of the level design team, so I contacted about it.

File: src/HUDTeam/DrawManagerImpl.java
Comment (LeeHyunWoo02) [reply src/HUDTeam/DrawManagerImpl.java:1]: Changing the name of the variable to bulletSpeed made it easier to understand. I don't think there's a problem with the code.

**Review by LeeHyunWoo02 — APPROVED** [review]
image

Code execution worked well without errors.
The updated code is also fine.

However, I am worried that the code will be safe as the button to enter the 2 player mode has disappeared.

File: src/entity/Ship.java (hunk 4 above)
Comment (kojesung) [reply src/entity/Ship.java:205]: Okay~

Comment (kojesung) [reply src/entity/Ship.java:205]: okay~

**Review by kojesung — APPROVED** [review]
image

I have verified the author-related comments and methods.

**Review by min5421 — APPROVED** [review]
bulletspeed

I checked if the attack speed function was changed to the bullet speed function.
Bullet speed value increases well when collecting bullet speed boost items.
The code is fine. Thank you.

**Pull request merged**